### PR TITLE
Implemented subscriber manager

### LINKED ISSUE #24: Unsubscribe topics from clients that disconnect
Right now inx-mqtt tracks the subscribed topics via `OnSubscribe` and `OnUnsubscribe` events of the MQTT broker.

If a client simply disconnects, without sending unsubscribe packages beforehand, inx-mqtt will always keep these topics as subscribed.

Instead of increasing the counter per subscribed topic, inx-mqtt should also remember the client IDs per topic.
By subscribing to `OnDisconnect`, inx-mqtt could then clean up the topics (or decrease the counter) that have not been unsubscribed by that client in a clean way.

### PR DESCRIPTION
**Pull request by oopsmonk**
This PR fixes #24 
The subscriber manager manages topics of subscribers based on connect/disconnect/subscribe/unsubscribe events

### CODE REVIEW

File: pkg/mqtt/subscriber_manager.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
+package mqtt
+
+import (
+	"sync"
+)
+
+type OnSubscribeH func(topic string, id string)
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: Call them `Func` at the end.
That fits better to the rest of our codebase.

File: core/mqtt/server.go [hunk 2]
```diff
@@ -190,6 +196,18 @@ func (s *Server) onUnsubscribeTopic(topic string) {
 	}
 }
 
+func (s *Server) onClientConnect(id string) {
+	// nothing to do at the moment
+}
+
+func (s *Server) onClientDisconnect(id string) {
+	// unsubscribe topics of the client which it might lost connection accidentally
+	s.LogDebug("onClientDisconnect")
```
Comment (muXxer) [inline core/mqtt/server.go]: Maybe also add the client ID in the Debug log like in the other logs.

File: core/mqtt/server.go [hunk 3]
```diff
@@ -190,6 +196,18 @@ func (s *Server) onUnsubscribeTopic(topic string) {
 	}
 }
 
+func (s *Server) onClientConnect(id string) {
+	// nothing to do at the moment
```
Comment (muXxer) [inline core/mqtt/server.go]: Also add a debug log here

File: core/mqtt/server.go [hunk 4]
```diff
@@ -123,7 +127,8 @@ func (s *Server) Run(ctx context.Context) {
 	s.MQTTBroker.Stop()
 }
 
-func (s *Server) onSubscribeTopic(ctx context.Context, topic string) {
+func (s *Server) onSubscribeTopic(ctx context.Context, id string, topic string) {
+	s.LogDebugf("%s subscriber: %s", id, topic)
```
Comment (muXxer) [inline core/mqtt/server.go]:
```suggestion
	s.LogDebugf("%s subscribed to topic %s", id, topic)
```

File: core/mqtt/server.go [hunk 5]
```diff
@@ -165,7 +170,8 @@ func (s *Server) onSubscribeTopic(ctx context.Context, topic string) {
 	}
 }
 
-func (s *Server) onUnsubscribeTopic(topic string) {
+func (s *Server) onUnsubscribeTopic(id string, topic string) {
+	s.LogDebugf("%s unsubscribe: %s", id, topic)
```
Comment (muXxer) [inline core/mqtt/server.go]:
```suggestion
	s.LogDebugf("%s unsubscribed from %s", id, topic)
```

File: pkg/mqtt/subscriber_manager.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+type OnUnsubscribeH func(topic string, id string)
+type OnConnectH func(id string)
+type OnDisconnectH func(id string)
+
+type subscriberManager struct {
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: I would call it `subscriptionManager`.

File: pkg/mqtt/subscriber_manager.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	// a map keeps client ID and topics
+	subscribers    map[string]map[string]string
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: Maybe call it `subscriptions` because it is not only subscribers, but also their topics.

File: pkg/mqtt/subscriber_manager.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	onUnsubscribe OnUnsubscribeH
+	onConnect     OnConnectH
+	onDisconnect  OnDisconnectH
+}
+
+func (s *subscriberManager) Connect(id string) {
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// add the client ID to map
+	s.subscribers[id] = make(map[string]string)
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: Maybe we should check if the client already exists?

File: pkg/mqtt/subscriber_manager.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+		s.onDisconnect(id)
+	}
+	// remove the client ID from map
+	delete(s.subscribers, id)
+}
+
+func (s *subscriberManager) Subscribe(id string, topicName string) {
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// add the topic to a crrosponding ID
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: typo

File: pkg/mqtt/subscriber_manager.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	s.subscribers[id][topicName] = topicName
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: better check if the client ID already exists, and create it if not.

File: pkg/mqtt/subscriber_manager.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+	if s.onSubscribe != nil {
+		s.onSubscribe(id, topicName)
+	}
+}
+
+func (s *subscriberManager) Unsubscribe(id string, topicName string) {
+	s.subscriberLock.Lock()
+	defer s.subscriberLock.Unlock()
+	// remove the topic from a crrosponding ID
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: typo

File: pkg/mqtt/subscriber_manager.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
+package mqtt
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: Maybe it would make sense to create test cases for the subscriptionManager.

File: pkg/mqtt/broker.go [hunk 13]
```diff
@@ -14,13 +14,13 @@ import (
 
 // Broker is a simple mqtt publisher abstraction.
 type Broker struct {
-	broker       *mqtt.Server
-	opts         *BrokerOptions
-	topicManager *topicManager
+	broker      *mqtt.Server
+	opts        *BrokerOptions
+	subMananger *subscriberManager
```
Comment (muXxer) [inline pkg/mqtt/broker.go]: `Mananger` is a typo.
I also would write a longer name, because "sub" doesn't ring a bell regarding subscriptions in my head :)

File: pkg/mqtt/subscriber_manager.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+	count := 0
+	for _, topics := range s.subscribers {
+		count += len(topics)
+	}
+	return count
+}
+
+// Returns topis of a subscriber
+func (s *subscriberManager) Topics(id string) map[string]string {
+	return s.subscribers[id]
```
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: I don't think it is a good idea to return the map, without holding the lock.
Maybe create an `ForEachTopic` instead, which gets a consumer function passed.

File: pkg/mqtt/topic_manager.go [hunk 15]
```diff
@@ -1,101 +0,0 @@
...
-	if t.onSubscribe != nil {
-		t.onSubscribe(topicName)
-	}
-}
-
-func (t *topicManager) Unsubscribe(topicName string) {
-	t.subscribedTopicsLock.Lock()
-	defer t.subscribedTopicsLock.Unlock()
-
-	count, has := t.subscribedTopics[topicName]
```
Comment (muXxer) [inline pkg/mqtt/topic_manager.go:43]: Why did you remove that whole logic with the counters?
This is essential in my opinion.

File: pkg/mqtt/topic_manager.go [hunk 16]
```diff
@@ -1,101 +0,0 @@
...
-	t.subscribedTopicsLock.RLock()
-	defer t.subscribedTopicsLock.RUnlock()
-
-	count, has := t.subscribedTopics[topicName]
-	return has && count > 0
-}
-
-// cleanupWithoutLocking recreates the subscribedTopics map to release memory for the garbage collector.
-func (t *topicManager) cleanupWithoutLocking() {
-	subscribedTopics := make(map[string]int)
```
Comment (muXxer) [inline pkg/mqtt/topic_manager.go:75]: Why did you remove the cleanup code?
Also essential.

File: pkg/mqtt/subscriber_manager.go (hunk 6 above)
Comment (muXxer) [inline pkg/mqtt/subscriber_manager.go]: Add the description what the subscriptionManager does (take the one from the topic manager and extend it)

**Review by muXxer — CHANGES_REQUESTED** [review]

File: pkg/mqtt/subscriber_manager.go (hunk 6 above)
Comment (oopsmonk) [reply pkg/mqtt/subscriber_manager.go]: it stores subscriber information for server handling subscriptions. the subscription management logic is implemented in the server.go, I'd like to keep it as `subscriberManager`.

File: pkg/mqtt/subscriber_manager.go (hunk 7 above)
Comment (oopsmonk) [reply pkg/mqtt/subscriber_manager.go]: call it subscribers is more specific, a subscriber has its own topics and a subscription is a collection of topics and subscribers, imo.

File: pkg/mqtt/subscriber_manager.go (hunk 8 above)
Comment (oopsmonk) [reply pkg/mqtt/subscriber_manager.go]: good catch!
it's more complex than just checking the existence, it depends on the mqtt client implementation, will figure out what's a proper way to deal with it later.

File: pkg/mqtt/subscriber_manager.go (hunk 14 above)
Comment (oopsmonk) [reply pkg/mqtt/subscriber_manager.go]: it's a get method which likely does not have race condition issues. the returned map is a copy.

File: pkg/mqtt/topic_manager.go (hunk 15 above)
Comment (oopsmonk) [reply pkg/mqtt/topic_manager.go:43]: the subscription counters are implemented in server.go. the topic manager is replaced by the subscriber manager, this counter is not necessary for subscriber manager.

File: pkg/mqtt/topic_manager.go (hunk 16 above)
Comment (oopsmonk) [reply pkg/mqtt/topic_manager.go:75]: ditto

File: pkg/mqtt/subscriber_manager.go (hunk 14 above)
Comment (alexsporn) [reply pkg/mqtt/subscriber_manager.go]: Map types are reference types, so there is no implicit copy here

File: pkg/mqtt/topic_manager.go (hunk 16 above)
Comment (muXxer) [reply pkg/mqtt/topic_manager.go:75]: This was added on purpose, because the map memory leaks, because the garbage collector never cleans it up.
This is a go specific problem.

Comment (oopsmonk) [reply pkg/mqtt/topic_manager.go:75]: I'd like to learn more about the GC issue with the map object, could you provide some materials? thanks in advance.

**Review by muXxer — APPROVED** [review]
